src/stan_math_backend/Cpp.ml

```diff
@@ -0,0 +1,732 @@
+  let matrix s = Matrix (s, -1, -1)
+
+  (** A [std::string]*)
+  let string = Type_literal "std::string"
+
+  let size_t = Type_literal "size_t"
+  let const_ref t = Const (Ref t)
+  let const_char_array i = Array (Const (Pointer (Type_literal "char")), i)
+end
+
+type operator =
+  | Multiply
+  | Divide
+  | Add
+  | Subtract
+  | Eq
+  | LEq
+  | GEq
+  | Lthn
+  | Gthn
+  | And
+  | Or
+[@@deriving sexp]
+
+type unary_op = PMinus | Incr [@@deriving sexp]
+
+type expr =
+  | Literal of string  (** printed as-is *)
+  | Var of identifier
+  | Parens of expr
+  | FunCall of identifier * type_ list * expr list
+  | MethodCall of expr * identifier * type_ list * expr list
+  | StaticMethodCall of type_ * identifier * type_ list * expr list
+  | Constructor of type_ * expr list  (** printed as [type(expr1, expr2, ...)]*)
+  | InitializerExpr of type_ * expr list
+      (** printed as [type{expr1, expr2, ...}]*)
+  | ArrayLiteral of expr list
+  | TernaryIf of expr * expr * expr
+  | Cast of type_ * expr
+  | Index of expr * expr
+  | New of string option * type_ * expr list
+  | Assign of expr * expr  (** NB: Not all exprs are valid lvalues! *)
+  | Unary of unary_op * expr
+  | StreamInsertion of expr * expr list  (** Corresponds to [operator<<] *)
+  | BinOp of expr * operator * expr
+[@@deriving sexp]
+
+module Exprs = struct
+  (** Some helper values and functions *)
+
+  (** Call a method on object, wrapping it in parentheses if
+      it is not a variable
+  *)
+  let method_call obj name templates args =
+    match obj with
+    | Var _ -> MethodCall (obj, name, templates, args)
+    | _ -> MethodCall (Parens obj, name, templates, args)
+
+  let to_var s = Var s
+
+  (** Turn an OCaml string into a quoted and escaped C++ string*)
+  let literal_string s = Literal ("\"" ^ Cpp_str.escaped s ^ "\"")
+
+  (** Equivalent to [std::vector<t>{e1,...,en}] *)
+  let std_vector_expr t es = InitializerExpr (Vector t, es)
+
+  let fun_call s es = FunCall (s, [], es)
+  let templated_fun_call s ts es = FunCall (s, ts, es)
+
+  (** Helper for [std::numeric_limits<double>::quiet_NaN()] *)
+  let quiet_NaN = fun_call "std::numeric_limits<double>::quiet_NaN" []
+
+  (** Helper for [std::numeric_limits<int>::min()] *)
+  let int_min = fun_call "std::numeric_limits<int>::min" []
+
+  (** Fold a list of expressions together with a C++ operator,
+      useful for e.g. adding a list of numbers. *)
+  let binop_list es ~f ~default : expr =
+    match es with
+    | [] -> default
+    | head :: rest -> List.fold ~init:head ~f:(fun a b -> Parens (f a b)) rest
+end
+
+module Expression_syntax = struct
+  (** Some operators to make streams and method calls look more
+      like the resultant C++ *)
+
+  include Exprs
+
+  (** A pun for the C++ [operator<<] *)
+  let ( << ) a b = StreamInsertion (a, b)
+
+  (** Method call: Call a no-argument method
+
+      E.g. [foo.bar()]
+  *)
+  let ( .@!() ) obj name = method_call obj name [] []
```

Comment:
[optional] Would be nice to give result types here if possible

Reply:
The return type of everything in Expression_syntax is `Cpp.expr`, so I'm not sure if it needs to be specified?